File: WorldFeed/Client/src/app/worldfeed/Client/Requesters/MessageRequester.ts
import {BaseRequester} from "../BaseRequester";
import {IMessageRequester} from "../../../core/Public/Client/Requesters/IMessageRequester";
import {IMessagesClientParametersValidator} from "../../../core/Core/Client/Validators/MessageClientParametersValidator";
import {ITwitterResult} from "../../../core/Core/Web/TwitterResult";
import {IMessageController} from "../../../core/Core/Controllers/IMessageController";
import {ITwitterClient} from "../../../core/Public/ITwitterClient";
import {ITwitterClientEvents} from "../../../core/Core/Events/TweetinviGlobalEvents";
import {IPublishMessageParameters} from "../../../core/Public/Parameters/MessageClient/PublishMessageParameters";
import {ICreateMessageDTO} from "../../../core/Public/Models/Interfaces/DTO/ICreateMessageDTO";
import {IDeleteMessageParameters} from "../../../core/Public/Parameters/MessageClient/DestroyMessageParameters";
import {IGetMessageDTO} from "../../../core/Public/Models/Interfaces/DTO/IGetMessageDTO";
import {IGetMessageParameters} from "../../../core/Public/Parameters/MessageClient/GetMessageParameters";
import {IGetMessagesParameters} from "../../../core/Public/Parameters/MessageClient/GetMessagesParameters";
import {ITwitterPageIterator} from "../../../core/Core/Iterators/TwitterPageIterator";
import {IMessageCursorQueryResultDTO} from "../../../core/Public/Models/Interfaces/DTO/QueryDTO/IMessageCursorQueryResultDTO";

export class MessageRequester extends BaseRequester implements IMessageRequester {
  private readonly _messageController: IMessageController;
  private readonly _messagesClientParametersValidator: IMessagesClientParametersValidator;

  constructor(client: ITwitterClient, messageController: IMessageController,
              messagesClientParametersValidator: IMessagesClientParametersValidator,
              twitterClientEvents: ITwitterClientEvents) {
    super(client, twitterClientEvents);
    this._messageController = messageController;
    this._messagesClientParametersValidator = messagesClientParametersValidator;
  }

  public publishMessageAsync(parameters: IPublishMessageParameters): Promise<ITwitterResult<ICreateMessageDTO>> {
    this._messagesClientParametersValidator.validate(parameters);
    return super.ExecuteRequestAsync(request => this._messageController.publishMessageAsync(parameters, request));
  }

  public destroyMessageAsync(parameters: IDeleteMessageParameters): Promise<ITwitterResult> {
    this._messagesClientParametersValidator.validate(parameters);
    return super.ExecuteRequestAsync(request => this._messageController.destroyMessageAsync(parameters, request));
  }

  public getMessageAsync(parameters: IGetMessageParameters): Promise<ITwitterResult<IGetMessageDTO>> {
    this._messagesClientParametersValidator.validate(parameters);
    return super.ExecuteRequestAsync(request => this._messageController.getMessageAsync(parameters, request));
  }

  public getMessagesIterator(parameters: IGetMessagesParameters): ITwitterPageIterator<ITwitterResult<IMessageCursorQueryResultDTO>> {
    this._messagesClientParametersValidator.validate(parameters);

    let request = super.TwitterClient.CreateRequest();
    request.executionContext.Converters = JsonQueryConverterRepository.Converters;
    return this._messageController.getMessagesIterator(parameters, request);
  }
}

File: Sharebook/Server/src/Services/Message/Sharebook.Message.API/Client/Requesters/MessageRequester.cs
namespace Sharebook.Message.Client.Requesters
{
    using System.Threading.Tasks;
    using Sharebook.Client;
    using Sharebook.Common.Client.Validators;
    using Sharebook.Common.Events;
    using Sharebook.Common.Iterators;
    using Sharebook.Common.JsonConverters;
    using Sharebook.Common.Public;
    using Sharebook.Common.Public.Client.Requesters;
    using Sharebook.Common.Public.Models.Interfaces.DTO;
    using Sharebook.Common.Public.Models.Interfaces.DTO.QueryDTO;
    using Sharebook.Common.Public.Parameters.MessageClient;
    using Sharebook.Common.Web;
    using Sharebook.Core.Controllers;

    public class MessageRequester : BaseRequester, IMessageRequester
    {
        private readonly IMessageController messageController;
        private readonly IMessagesClientParametersValidator messagesClientParametersValidator;

        public MessageRequester(
            ITwitterClient client,
            IMessageController messageController,
            IMessagesClientParametersValidator messagesClientParametersValidator,
            ITwitterClientEvents twitterClientEvents)
            : base(client, twitterClientEvents)
        {
            this.messageController = messageController;
            this.messagesClientParametersValidator = messagesClientParametersValidator;
        }

        public Task<ITwitterResult<ICreateMessageDTO>> PublishMessageAsync(IPublishMessageParameters parameters)
        {
            this.messagesClientParametersValidator.Validate(parameters);
            return ExecuteRequestAsync(request => this.messageController.PublishMessageAsync(parameters, request));
        }

        public Task<ITwitterResult> DestroyMessageAsync(IDeleteMessageParameters parameters)
        {
            this.messagesClientParametersValidator.Validate(parameters);
            return ExecuteRequestAsync(request => this.messageController.DestroyMessageAsync(parameters, request));
        }

        public Task<ITwitterResult<IGetMessageDTO>> GetMessageAsync(IGetMessageParameters parameters)
        {
            this.messagesClientParametersValidator.Validate(parameters);
            return ExecuteRequestAsync(request => this.messageController.GetMessageAsync(parameters, request));
        }

        public ITwitterPageIterator<ITwitterResult<IMessageCursorQueryResultDTO>> GetMessagesIterator(IGetMessagesParameters parameters)
        {
            this.messagesClientParametersValidator.Validate(parameters);

            var request = TwitterClient.CreateRequest();
            request.ExecutionContext.Converters = JsonQueryConverterRepository.Converters;
            return this.messageController.GetMessagesIterator(parameters, request);
        }
    }
}